zio-http/shared/src/main/scala/zio/http/Route.scala

```diff
@@ -118,6 +118,33 @@ sealed trait Route[-Env, +Err] { self =>
         Handled(rpm.routePattern, handler2, location)
     }
 
+  /**
+   * Effectfully peeks at the unhandled failure of this Route.
+   */
+  final def tapErrorZIO[Err1 >: Err](
+    f: Err => ZIO[Any, Err1, Any],
+  )(implicit trace: Trace): Route[Env, Err1] =
+    self match {
+      case Provided(route, env)                        => Provided(route.tapErrorZIO(f), env)
+      case Augmented(route, aspect)                    => Augmented(route.tapErrorZIO(f), aspect)
+      case Handled(routePattern, handler, location)    => Handled(routePattern, handler, location)
+      case Unhandled(rpm, handler, zippable, location) => Unhandled(rpm, handler.tapErrorZIO(f), zippable, location)
+    }
+
+  /**
+   * Effectfully peeks at the unhandled failure cause of this Route.
+   */
+  final def tapErrorCauseZIO[Err1 >: Err](
+    f: Cause[Err] => ZIO[Any, Err1, Any],
+  )(implicit trace: Trace): Route[Env, Err1] =
+    self match {
+      case Provided(route, env)                        => Provided(route.tapErrorCauseZIO(f), env)
+      case Augmented(route, aspect)                    => Augmented(route.tapErrorCauseZIO(f), aspect)
+      case Handled(routePattern, handler, location)    => Handled(routePattern, handler, location)
```

Comment:
iirc, ensures handled that the typed error is handled. But a cause (defect) might still happen. So I think there should be a tap on the handler here.

Reply:
That doesn't seem to work as the `Handled.handler`'s failure type is `Response` whereas `f`'s failure type is `Err` which is not necessarily a response.

Reply:
@tjarvstrand This does need to be fixed, and then will be good to merge!

Reply:
Sorry, busy week! :/

I've been struggling a bit with the failure case of `f`. Does something like this make sense?

```
/**
   * Effectfully peeks at the unhandled failure cause of this Route.
   */
  final def tapErrorCauseZIO[Err1 >: Err](
    f: Cause[Err] => ZIO[Any, Err1, Any],
  )(implicit trace: Trace): Route[Env, Err1] =
    self match {
      case Provided(route, env)                        => Provided(route.tapErrorCauseZIO(f), env)
      case Augmented(route, aspect)                    => Augmented(route.tapErrorCauseZIO(f), aspect)
      case Handled(routePattern, handler, location)    =>
        Handled(
          routePattern,
          handler.tapErrorCauseZIO {
            case cause0: Cause[Err] => f(cause0).catchAllCause(cause => ZIO.succeed(Response.fromCause(cause)))
            case _                  => ZIO.unit
          },
          location,
        )
      case Unhandled(rpm, handler, zippable, location) =>
        Unhandled(rpm, handler.tapErrorCauseZIO(f), zippable, location)
    }
```

Reply:
Sorry, busy week! :/

I've been struggling a bit with the error case of `f` which, unless I'm mistaken, must be a `Response`. Does something like this make sense?
```
/**
   * Effectfully peeks at the unhandled failure cause of this Route.
   */
  final def tapErrorCauseZIO[Err1 >: Err](
    f: Cause[Err] => ZIO[Any, Err1, Any],
  )(implicit trace: Trace): Route[Env, Err1] =
    self match {
      case Provided(route, env)                        => Provided(route.tapErrorCauseZIO(f), env)
      case Augmented(route, aspect)                    => Augmented(route.tapErrorCauseZIO(f), aspect)
      case Handled(routePattern, handler, location)    =>
        Handled(
          routePattern,
          handler.tapErrorCauseZIO {
            case cause0: Cause[Err] => f(cause0).catchAllCause(cause => ZIO.fail(Response.fromCause(cause)))
            case _                  => ZIO.unit
          },
          location,
        )
      case Unhandled(rpm, handler, zippable, location) =>
        Unhandled(rpm, handler.tapErrorCauseZIO(f), zippable, location)
    }
```

Reply:
When you get `cause0`, do `errorOrCause`, and if it's an error, you have to refail that, but if it's a `Cause[Nothing]`, you can feed it to `f`.

Reply:
Maybe I'm slow, but shouldn't `handler.tapErrorCauseZIO` leave the error channel of the original effect intact so that I don't have to refail it?

Reply:
Ah yes, of course! Forgot the point of this function, which is tapping not recovering. 🤦‍♂️

Reply:
👍 

Tried to move too fast and now the build is failing. I'll have another look at this as soon as I can.

Reply:
I'm afraid I think I'm in need of some assistance.

There are warnings in Scala 3 (which I have actually not used yet):
```
[warn] -- Unchecked Warning: /Users/tjarvstrand/src/zio-http/zio-http/shared/src/main/scala/zio/http/Route.scala:134:17
[warn] 134 |            case err: Fail[Err] => f(err.value).catchAllCause(cause => ZIO.fail(Response.fromCause(cause)))
[warn]     |                 ^
[warn]     |the type test for zio.Cause.Fail[Err] cannot be checked at runtime because its type arguments can't be determined from zio.Cause[zio.http.Response]
```

I'm assuming this is related to type erasure somehow, but I'm not sure why there's no warning in Scala 2. Any tips on how to proceed?

Reply:
@tjarvstrand You can use the own methods of `Cause` to determine whether it's a failure or a defect.

You can refer to [the implementation of `ZIO#tapDeffect` and `ZIO#tapError`](https://github.com/zio/zio/blob/v2.1.3/core/shared/src/main/scala/zio/ZIO.scala#L2034-L2065), and there might be a more convenient method to fold defect and failure separately.

Reply:
@guersam Thanks, AFAICT the issue here is checking whether the `f`s input (`Cause[Err]`) is a subtype of `Cause[Response]` or not, and only call it then

Reply:
@tjarvstrand What I meant originally was we don't need pattern matching to determine if it's `Cause[Nothing]` or not, according to John's previous comment.

> You can catch the Cause[Err], and if it contains an Err, then you rethrow it. But if it's really a Cause[Nothing], i.e. pure defect, without an Err inside it, then you can allow that to be caught by f, since f can handle those (Cause[Nothing] is a subtype of Cause[Err]).

But now we want to let `f` handle failure cases as well when `Err =:= Response`, right?

I think there are two options:
1. Extract the possible failure with something like `Cause#failureOrCause`, and do runtime type check for the failure case using `.isInstanceOf[Response]`. It's the simplest solution.
2. Add an implicit evidence like this:

 ```
sealed trait IsResponse[-A] { def isResponse: Boolean }
object IsResponse {
  implicit val response: IsResponse[Response] = new IsResponse[Response] { isResponse = true }
  implicit def other: IsResponse[A] = new IsResponse[A] { isResponse = false }
}
```

The name `IsResponse` might cause confusion with `Handler.IsRequest`, which has the similar purpose but it's for compile time check only.

Reply:
By the way, `handler` is a handler that produces handler ([code](https://github.com/zio/zio-http/blob/v3.0.0-RC9/zio-http/shared/src/main/scala/zio/http/Route.scala#L460)), so checking `handler`'s failure won't achieve what you want. You need `handler.map(_.tapErrorCauseZIO(...))` to catch the `Response` failure.

Reply:
Thanks a lot for your comments!

If I understand it correctly, both alternative 2 and 3 would require a call to `asInstanceOf`, since neither of them provide a type level proof that `Err <: Response`. Is that correct?

> By the way, Handled.handler is a handler that produces another handler ([code](https://github.com/zio/zio-http/blob/v3.0.0-RC9/zio-http/shared/src/main/scala/zio/http/Route.scala#L460)), so checking handler's failure won't achieve what you want. You need handler.map(_.tapErrorCauseZIO(...)) to catch the Response failure.

Aha, I missed that! Thanks!

Reply:
BTW, in other parts of `Route.scala`, I see things like this:
```
handler.map { handler =>
  Handler.fromFunctionHandler[Request] { (req: Request) =>
    handler.mapErrorCause(c => f(req, c.asInstanceOf[Cause[Nothing]]))
  }
}
```

Is this really safe? To me it looks like there has perhaps been a mixup, since the outer `handler` is infallible, but the inner one isn't.

Reply:
You cannot summon `ClassTag[Err]` directly, but you can introduce new type variable `Err1`.

Refer to [ZIO#refinedToOrDie](https://github.com/zio/zio/blob/v2.1.5/core/shared/src/main/scala/zio/ZIO.scala#L5320-L5321):

```scala
def refineToOrDie[E1 <: E: ClassTag](implicit ev: CanFail[E], trace: Trace): ZIO[R, E1, A] =
  self.refineOrDie { case e: E1 => e }
```